File: sdk/python/src/zincha/bincode.py
"""Small bincode writer for the Rust transaction wire format."""

from __future__ import annotations

import struct
from typing import Callable, Iterable, Optional, TypeVar, Union

BigNumberish = Union[int, str]
T = TypeVar("T")


def as_u64(value: BigNumberish, field: str = "value") -> int:
    parsed = int(value)
    if parsed < 0 or parsed > 0xFFFF_FFFF_FFFF_FFFF:
        raise ValueError("%s must fit in unsigned 64 bits" % field)
    return parsed


def as_u32(value: int, field: str = "value") -> int:
    if int(value) != value or value < 0 or value > 0xFFFF_FFFF:
        raise ValueError("%s must fit in unsigned 32 bits" % field)
    return int(value)


def as_u8(value: int, field: str = "value") -> int:
    if int(value) != value or value < 0 or value > 0xFF:
        raise ValueError("%s must fit in unsigned 8 bits" % field)
    return int(value)


class BincodeWriter:
    def __init__(self) -> None:
        self._chunks = []

    def write_u8(self, value: int) -> None:
        self._chunks.append(bytes([as_u8(value)]))

    def write_u32(self, value: int) -> None:
        self._chunks.append(as_u32(value).to_bytes(4, "little"))

    def write_u64(self, value: BigNumberish) -> None:
        self._chunks.append(as_u64(value).to_bytes(8, "little"))

    def write_f32(self, value: float) -> None:
        self._chunks.append(struct.pack("<f", float(value)))

    def write_f64(self, value: float) -> None:
        self._chunks.append(struct.pack("<d", float(value)))

    def write_string(self, value: str) -> None:
        self.write_bytes(value.encode("utf-8"))

    def write_bytes(self, value: bytes) -> None:
        self.write_u64(len(value))
        self._chunks.append(bytes(value))

    def write_raw(self, value: bytes) -> None:
        self._chunks.append(bytes(value))

    def write_option(
        self,
        value: Optional[T],
        encode: Callable[["BincodeWriter", T], None],
    ) -> None:
        """bincode Option<T>: 0u8 (None) or 1u8 + T (Some)."""
        if value is None:
            self.write_u8(0)
        else:
            self.write_u8(1)
            encode(self, value)

    def write_vec(
        self,
        values: Iterable[T],
        encode: Callable[["BincodeWriter", T], None],
    ) -> None:
        """bincode Vec<T>: u64 length + each element."""
        materialized = list(values)
        self.write_u64(len(materialized))
        for value in materialized:
            encode(self, value)

    def finish(self) -> bytes:
        return b"".join(self._chunks)

File: sdk/typescript/src/bincode.ts
import type { BigNumberish } from "./types.ts";

export function asU64(value: BigNumberish, field = "value"): bigint {
  const parsed = typeof value === "bigint" ? value : BigInt(value);
  if (parsed < 0n || parsed > 0xffff_ffff_ffff_ffffn) {
    throw new Error(`${field} must fit in unsigned 64 bits`);
  }
  return parsed;
}

export function asU32(value: number, field = "value"): number {
  if (!Number.isInteger(value) || value < 0 || value > 0xffff_ffff) {
    throw new Error(`${field} must fit in unsigned 32 bits`);
  }
  return value;
}

export function asU16(value: number, field = "value"): number {
  if (!Number.isInteger(value) || value < 0 || value > 0xffff) {
    throw new Error(`${field} must fit in unsigned 16 bits`);
  }
  return value;
}

export class BincodeWriter {
  private readonly chunks: Uint8Array[] = [];

  writeU8(value: number): void {
    if (!Number.isInteger(value) || value < 0 || value > 0xff) {
      throw new Error("value must fit in unsigned 8 bits");
    }
    this.chunks.push(new Uint8Array([value]));
  }

  writeU16(value: number): void {
    const out = new Uint8Array(2);
    new DataView(out.buffer).setUint16(0, asU16(value), true);
    this.chunks.push(out);
  }

  writeU32(value: number): void {
    const out = new Uint8Array(4);
    new DataView(out.buffer).setUint32(0, asU32(value), true);
    this.chunks.push(out);
  }

  writeU64(value: BigNumberish): void {
    const out = new Uint8Array(8);
    new DataView(out.buffer).setBigUint64(0, asU64(value), true);
    this.chunks.push(out);
  }

  writeF32(value: number): void {
    const out = new Uint8Array(4);
    new DataView(out.buffer).setFloat32(0, value, true);
    this.chunks.push(out);
  }

  writeF64(value: number): void {
    const out = new Uint8Array(8);
    new DataView(out.buffer).setFloat64(0, value, true);
    this.chunks.push(out);
  }

  writeString(value: string): void {
    this.writeBytes(new TextEncoder().encode(value));
  }

  writeBytes(value: Uint8Array): void {
    this.writeU64(value.length);
    this.chunks.push(value);
  }

  writeRaw(value: Uint8Array): void {
    this.chunks.push(value);
  }

  /** bincode Option<T>: 0u8 (None) or 1u8 + T (Some). */
  writeOption<T>(value: T | null | undefined, encode: (writer: BincodeWriter, value: T) => void): void {
    if (value === null || value === undefined) {
      this.writeU8(0);
    } else {
      this.writeU8(1);
      encode(this, value);
    }
  }

  /** bincode Vec<T>: u64 length + each element. */
  writeVec<T>(values: readonly T[], encode: (writer: BincodeWriter, value: T) => void): void {
    this.writeU64(values.length);
    for (const value of values) {
      encode(this, value);
    }
  }

  finish(): Uint8Array {
    const total = this.chunks.reduce((sum, chunk) => sum + chunk.length, 0);
    const out = new Uint8Array(total);
    let offset = 0;
    for (const chunk of this.chunks) {
      out.set(chunk, offset);
      offset += chunk.length;
    }
    return out;
  }
}
